### DISCUSSION TITLE
Default values for ResourceDependency

### DISCUSSION THREAD

**Discussion by daviskirk** [dagster-io/dagster#19447] [Q&A]
How do I supply a default value to a ConfigurableResource class.

Example:

```
from dagster import ConfigurableResource, ResourceDependency

class MyResource(ConfigurableResource):
    initial_path: ResourceDependency[Path] = Path(".")  # mypy complains here because it is expecting a `ResourceDepdendency` object
```

**Comment by garethbrickman** [comment]
Hey @daviskirk I let our LLM Support Bot take a pass at this one below. Please let me know if its response proves helpful! 

To supply a default value to a `ConfigurableResource` class in Dagster, you would typically define the default value for the configuration directly in the class definition. However, it's important to note that `ResourceDependency` is not a type that you would use to set a default value for a configuration field. Instead, you would define the configuration field and its default value without using `ResourceDependency`.

Here's an example of how you might define a `ConfigurableResource` with a default value for a configuration field:

```python
from dagster import ConfigurableResource

class MyResource(ConfigurableResource):
    initial_path: str = "default/path"
```

In this example, `initial_path` is a configuration field of type `str`, and it has a default value of `"default/path"`. When you use this resource in a job or asset, if you do not provide a value for `initial_path` in the run configuration, Dagster will use the default value specified in the class definition.

If you want to use a `Path` object instead of a string, you would need to handle the conversion from the string path to a `Path` object within the resource's methods or its initialization logic. Here's an example of how you might do that:

```python
from pathlib import Path
from dagster import ConfigurableResource

class MyResource(ConfigurableResource):
    initial_path: str = "default/path"

    def __init__(self, initial_path: str):
        self.path = Path(initial_path)

    # ... other methods that use self.path ...
```

In this revised example, the `initial_path` configuration field is still a string, but the `__init__` method of the resource converts it to a `Path` object, which is then stored in an instance variable `self.path` for use in other methods of the resource.

Remember that the `ResourceDependency` type is not used for setting default values; it is used to express dependencies between resources. If you need to define dependencies between resources, you would do so in the job or asset definition, not in the `ConfigurableResource` class itself.

**Reply by daviskirk** [reply]
Perhaps I should specify my problem in more detail:

The I want the resource:

```
class MyResource(ConfigurableResource):
    initial_path: ResourceDependency[Path] = Path(".")  # mypy complains here because it is expecting a `ResourceDepdendency` object
```

To be able to depend on a different resource that provides the initial_path for each run context.
An example might be to generate a new temporary path for each run which is used across different resources within that run.
However, in other cases, this dynamic resource is not needed and I just want the default path to fall back to the current directory (i.e. ".") in these cases (or be able to provide a static path).
